File: packages/cli/src/test/helpers/pty.ts
import { chmodSync, realpathSync, statSync } from "node:fs";
import { createRequire } from "node:module";
import { dirname, join } from "node:path";
import { stripVTControlCharacters } from "node:util";

const require = createRequire(import.meta.url);

let isPtyFixed = false;

/**
 * On macOS, node-pty's `spawn-helper` binary sometimes lacks the executable bit
 * after installation, leading to "posix_spawnp failed" errors.
 * This function programmatically ensures the executable bit is set for the helper
 * binaries on Darwin platforms.
 */
const fixPtyPermissions = () => {
  if (isPtyFixed || process.platform !== "darwin") {
    return;
  }

  try {
    const ptyRoot = dirname(require.resolve("node-pty/package.json"));
    const helpers = [
      join(ptyRoot, "prebuilds", "darwin-arm64", "spawn-helper"),
      join(ptyRoot, "prebuilds", "darwin-x64", "spawn-helper"),
    ];

    for (const helper of helpers) {
      try {
        const stats = statSync(helper);
        if ((stats.mode & 0o111) !== 0o111) {
          chmodSync(helper, stats.mode | 0o111);
        }
      } catch {
        // Ignore
      }
    }
  } catch {
    // Ignore
  }

  isPtyFixed = true;
};

const applyBackspaces = (value: string) => {
  let result = "";

  for (const character of value) {
    if (character === "\b") {
      result = result.slice(0, -1);
      continue;
    }

    result += character;
  }

  return result;
};

const normalizeTerminalOutput = (value: string) => {
  return applyBackspaces(stripVTControlCharacters(value))
    .replace(/\r\n/gu, "\n")
    .replace(/\r/gu, "\n");
};

export type PtySession = Readonly<{
  clearOutput: () => void;
  close: () => void;
  getOutput: () => string;
  waitFor: (pattern: RegExp | string, timeoutMs?: number) => Promise<string>;
  waitForOutput: (
    predicate: (output: string) => boolean,
    timeoutMs?: number,
  ) => Promise<string>;
  waitForExit: (timeoutMs?: number) => Promise<void>;
  write: (value: string) => void;
}>;

export const createPtyOutputWaiter = (options: {
  getOutput: () => string;
  onOutput: (listener: () => void) => { dispose: () => void };
}) => {
  const waitForOutput = (
    predicate: (output: string) => boolean,
    timeoutMs = 10_000,
  ) => {
    return new Promise<string>((resolve, reject) => {
      const currentOutput = options.getOutput();

      if (predicate(currentOutput)) {
        resolve(currentOutput);
        return;
      }

      let outputSubscription: { dispose: () => void } | undefined;
      const timeout = setTimeout(() => {
        cleanup();
        reject(
          new Error(
            `Timed out waiting for terminal output predicate.\n\n${options.getOutput()}`,
          ),
        );
      }, timeoutMs);

      const cleanup = () => {
        clearTimeout(timeout);
        outputSubscription?.dispose();
      };

      outputSubscription = options.onOutput(() => {
        const nextOutput = options.getOutput();

        if (!predicate(nextOutput)) {
          return;
        }

        cleanup();
        resolve(nextOutput);
      });
    });
  };

  return { waitForOutput };
};

export const createPtySession = (options: {
  args?: readonly string[];
  cwd: string;
  env?: Readonly<Record<string, string | undefined>>;
  file: string;
}): PtySession => {
  fixPtyPermissions();

  const pty = require("node-pty") as typeof import("node-pty");

  const cleanEnv: Record<string, string> = {};

  for (const [key, value] of Object.entries({
    ...process.env,
    ...options.env,
  })) {
    if (value !== undefined) {
      cleanEnv[key] = value;
    }
  }

  const terminal = pty.spawn(options.file, [...(options.args ?? [])], {
    cols: 120,
    // On macOS, the temporary directory /var is a symlink to /private/var.
    // node-pty (posix_spawn) can fail if the CWD is a symlinked path.
    // We resolve it to the real path to ensure stability.
    cwd: realpathSync(options.cwd),
    env: {
      ...cleanEnv,
      TERM: "xterm-256color",
    },
    name: "xterm-256color",
    rows: 40,
  });

  let output = "";
  let exitCode: number | undefined;

  terminal.onData((chunk) => {
    output += chunk;
  });
  terminal.onExit((event) => {
    exitCode = event.exitCode;
  });

  const getOutput = () => {
    return normalizeTerminalOutput(output);
  };
  const outputWaiter = createPtyOutputWaiter({
    getOutput,
    onOutput: (listener) => terminal.onData(listener),
  });

  return {
    clearOutput: () => {
      output = "";
    },
    close: () => {
      terminal.kill();
    },
    getOutput,
    waitFor: (pattern, timeoutMs = 10_000) => {
      return new Promise((resolve, reject) => {
        outputWaiter
          .waitForOutput((normalizedOutput) => {
            return typeof pattern === "string"
              ? normalizedOutput.includes(pattern)
              : pattern.test(normalizedOutput);
          }, timeoutMs)
          .then(resolve, (error: unknown) => {
            reject(
              new Error(
                `Timed out waiting for terminal output matching ${String(pattern)}.\n\n${getOutput()}`,
                { cause: error },
              ),
            );
          });
      });
    },
    waitForOutput: outputWaiter.waitForOutput,
    waitForExit: (timeoutMs = 10_000) => {
      return new Promise((resolve, reject) => {
        if (exitCode !== undefined) {
          resolve();
          return;
        }

        let subscription: { dispose: () => void } | undefined;
        const timeout = setTimeout(() => {
          subscription?.dispose();
          reject(new Error("Timed out waiting for terminal exit."));
        }, timeoutMs);
        subscription = terminal.onExit(() => {
          clearTimeout(timeout);
          subscription?.dispose();
          resolve();
        });
      });
    },
    write: (value) => {
      terminal.write(value);
    },
  };
};

File: packages/cli/test/helpers/pty.dart
// Dart port of `packages/cli/src/test/helpers/pty.ts` plus
// `packages/cli/src/test/helpers/shell-availability.ts`.
//
// The TS helper drives interactive shells through node-pty. Dart has no pty
// binding, so this port wraps the POSIX `script` utility, which allocates a
// pty for the spawned command and mirrors the pty output on its own stdout:
//
// - Linux (util-linux): `script -qefc "<command>" /dev/null`
// - macOS (BSD):        `script -q /dev/null <command> <args...>`
//
// Input written to the session goes to `script`'s stdin, which forwards it to
// the pty master (so the shell under test sees a real tty). node-pty pinned
// the terminal to 120x40; `script` gets no window size from a piped stdin, so
// COLUMNS/LINES are exported as a hint instead.
//
// Windows has no `script` utility; [startPtySession] throws there. The pty
// e2e suites guard every test body with `if (Platform.isWindows) return;`
// (the TS suites' `process.platform !== "win32"` skip), so on Windows they
// compile but never spawn a session.
//
// The macOS node-pty `spawn-helper` chmod workaround from `pty.ts` has no
// Dart equivalent and is intentionally dropped.

import 'dart:async';
import 'dart:convert';
import 'dart:io';

// --- Port of `shell-availability.ts` ---------------------------------------

String? _probeShellPath(String shell) {
  final lookupCommand = Platform.isWindows ? 'where' : 'which';
  final isUnsupportedWindowsShell = Platform.isWindows && shell == 'bash';
  final ProcessResult result;

  try {
    result = Process.runSync(
      lookupCommand,
      [shell],
      stdoutEncoding: utf8,
      stderrEncoding: utf8,
    );
  } on ProcessException {
    return null;
  }

  // `execFileSync` throws on a non-zero exit code; the TS catch returns
  // undefined.
  if (result.exitCode != 0) {
    return null;
  }

  final unsupportedWindowsBashPattern = RegExp(
    r'\\(?:Windows\\System32|Microsoft\\WindowsApps)\\bash\.exe$',
    caseSensitive: false,
  );

  for (final rawLine in (result.stdout as String).split(RegExp(r'\r?\n'))) {
    final line = rawLine.trim();

    if (line.isEmpty) {
      continue;
    }

    if (isUnsupportedWindowsShell &&
        unsupportedWindowsBashPattern.hasMatch(line)) {
      continue;
    }

    return line;
  }

  return null;
}

final String? bashPath = _probeShellPath('bash');
final String? fishPath = _probeShellPath('fish');
final String? powerShellPath =
    _probeShellPath('pwsh') ?? _probeShellPath('powershell');
final String? zshPath = _probeShellPath('zsh');

final bool isBashAvailable = bashPath != null;
final bool isFishAvailable = fishPath != null;
final bool isPowerShellAvailable = powerShellPath != null;
final bool isZshAvailable = zshPath != null;

// --- Port of `pty.ts` -------------------------------------------------------

/// Mirror of the TS `applyBackspaces`.
String applyBackspaces(String value) {
  final result = <String>[];

  for (final rune in value.runes) {
    if (rune == 0x08) {
      if (result.isNotEmpty) {
        result.removeLast();
      }
      continue;
    }

    result.add(String.fromCharCode(rune));
  }

  return result.join();
}

/// Mirror of Node's `util.stripVTControlCharacters` (the ansi-regex pattern).
final String _escapeCharacter = String.fromCharCode(0x1B);
final String _csiCharacter = String.fromCharCode(0x9B);
final String _bellCharacter = String.fromCharCode(0x07);
final String _stringTerminatorCharacter = String.fromCharCode(0x9C);

final RegExp _vtControlPattern = RegExp(
  '[$_escapeCharacter$_csiCharacter]'
  r'[\[\]()#;?]*'
  r'(?:(?:(?:(?:;[-a-zA-Z\d/#&.:=?%@~_]+)*'
  r'|[a-zA-Z\d]+(?:;[-a-zA-Z\d/#&.:=?%@~_]*)*)?'
  '(?:$_bellCharacter|$_escapeCharacter'
  r'\\'
  '|$_stringTerminatorCharacter))'
  r'|(?:(?:\d{1,4}(?:;\d{0,4})*)?[\dA-PR-TZcf-nq-uy=><~]))',
);

/// Mirror of Node's `stripVTControlCharacters` used by the TS helper.
String stripVtControlCharacters(String value) {
  return value.replaceAll(_vtControlPattern, '');
}

/// Mirror of the TS `normalizeTerminalOutput`.
String normalizeTerminalOutput(String value) {
  return applyBackspaces(
    stripVtControlCharacters(value),
  ).replaceAll('\r\n', '\n').replaceAll('\r', '\n');
}

String _posixShellQuote(String value) {
  return "'${value.replaceAll("'", "'\\''")}'";
}

/// Mirror of the TS `PtySession` interface (`waitFor` and `waitForOutput`
/// take a [Duration] instead of milliseconds).
class PtySession {
  PtySession._(this._process) {
    const decoder = Utf8Decoder(allowMalformed: true);

    _process.stdout.transform(decoder).listen(_append);
    _process.stderr.transform(decoder).listen(_append);
    // The process may exit without reading stdin; swallow the broken-pipe
    // error instead of surfacing an unhandled async exception.
    unawaited(_process.stdin.done.then((_) {}, onError: (Object _) {}));
    unawaited(
      _process.exitCode.then((code) {
        _exitCode = code;
      }),
    );
  }

  final Process _process;
  final StreamController<void> _onOutput = StreamController<void>.broadcast();
  String _raw = '';
  int? _exitCode;

  void _append(String chunk) {
    _raw += chunk;
    _onOutput.add(null);
  }

  /// Mirror of `clearOutput`.
  void clearOutput() {
    _raw = '';
  }

  /// Mirror of `close` (node-pty `terminal.kill()`): kills the `script`
  /// wrapper, which closes the pty master and hangs up the shell inside it.
  void close() {
    unawaited(_process.stdin.close().then((_) {}, onError: (Object _) {}));
    _process.kill();
  }

  /// Mirror of `getOutput` (normalized terminal output).
  String getOutput() {
    return normalizeTerminalOutput(_raw);
  }

  /// Mirror of `waitForOutput`.
  Future<String> waitForOutput(
    bool Function(String output) predicate, [
    Duration timeout = const Duration(seconds: 10),
  ]) {
    final current = getOutput();

    if (predicate(current)) {
      return Future.value(current);
    }

    final completer = Completer<String>();
    late final StreamSubscription<void> subscription;
    final timer = Timer(timeout, () {
      unawaited(subscription.cancel());
      if (!completer.isCompleted) {
        completer.completeError(
          StateError(
            'Timed out waiting for terminal output predicate.'
            '\n\n${getOutput()}',
          ),
        );
      }
    });

    subscription = _onOutput.stream.listen((_) {
      final next = getOutput();

      if (!predicate(next)) {
        return;
      }

      timer.cancel();
      unawaited(subscription.cancel());
      if (!completer.isCompleted) {
        completer.complete(next);
      }
    });

    return completer.future;
  }

  /// Mirror of `waitFor` (accepts a plain [String] or a [RegExp]).
  Future<String> waitFor(
    Pattern pattern, [
    Duration timeout = const Duration(seconds: 10),
  ]) async {
    try {
      return await waitForOutput((output) => output.contains(pattern), timeout);
    } on StateError {
      throw StateError(
        'Timed out waiting for terminal output matching $pattern.'
        '\n\n${getOutput()}',
      );
    }
  }

  /// Mirror of `waitForExit`.
  Future<void> waitForExit([
    Duration timeout = const Duration(seconds: 10),
  ]) async {
    if (_exitCode != null) {
      return;
    }

    try {
      await _process.exitCode.timeout(timeout);
    } on TimeoutException {
      throw StateError('Timed out waiting for terminal exit.');
    }
  }

  /// Mirror of `write` (sent to the pty through `script`'s stdin).
  void write(String value) {
    try {
      _process.stdin.add(utf8.encode(value));
    } on StateError {
      // Writing after the wrapper exited is a no-op, like node-pty.
    }
  }
}

/// Mirror of the TS `createPtySession` (async because Dart process spawning
/// is): starts [file] with [args] inside a pty via the platform's `script`
/// utility. [env] entries are layered over the parent environment; `null`
/// values drop the variable (the TS `undefined` cleanup).
Future<PtySession> startPtySession({
  List<String> args = const [],
  required String cwd,
  Map<String, String?> env = const {},
  required String file,
}) async {
  final String executable;
  final List<String> scriptArgs;

  if (Platform.isLinux) {
    // util-linux: -q quiet, -e return child exit code, -f flush, -c command.
    executable = 'script';
    scriptArgs = [
      '-qefc',
      [file, ...args].map(_posixShellQuote).join(' '),
      '/dev/null',
    ];
  } else if (Platform.isMacOS) {
    // BSD script: the command and its arguments are positional.
    executable = 'script';
    scriptArgs = ['-q', '/dev/null', file, ...args];
  } else {
    throw UnsupportedError(
      'PTY sessions require the POSIX `script` utility; '
      'unsupported on ${Platform.operatingSystem}.',
    );
  }

  final merged = <String, String?>{...Platform.environment, ...env};
  final environment = <String, String>{};

  for (final entry in merged.entries) {
    final value = entry.value;

    if (value != null) {
      environment[entry.key] = value;
    }
  }

  environment['TERM'] = 'xterm-256color';
  // node-pty spawned with cols: 120, rows: 40; `script` cannot inherit a
  // window size from a piped stdin, so hint the geometry through the
  // environment instead.
  environment['COLUMNS'] = '120';
  environment['LINES'] = '40';

  final process = await Process.start(
    executable,
    scriptArgs,
    // On macOS the temporary directory /var is a symlink to /private/var;
    // resolve it like the TS helper does for node-pty.
    workingDirectory: await Directory(cwd).resolveSymbolicLinks(),
    environment: environment,
    includeParentEnvironment: false,
  );

  return PtySession._(process);
}
